### PR TITLE
Add k8s service resource

### PR DESCRIPTION
**Pull request by rossf7**
Towards giantswarm/giantswarm#2023

Adds the service resource for guest cluster scraping. I'll enable once the endpoint resource is ready.

### CODE REVIEW

File: service/resource/namespacev2/resource.go [hunk 1]
```diff
@@ -13,14 +13,14 @@ const (
 	Name = "namespacev2"
 )
 
-// Config represents the configuration used to create a new deployment resource.
+// Config represents the configuration used to create a new namespace resource.
```
Comment (fgimenez) [inline service/resource/namespacev2/resource.go]: <3

File: service/resource/servicev2/create_test.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	"github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/micrologger/microloggertest"
+	apismetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+	apiv1 "k8s.io/client-go/pkg/api/v1"
+)
+
+func Test_Resource_Service_newCreateChange(t *testing.T) {
+	testCases := []struct {
+		Obj             interface{}
```
Comment (fgimenez) [inline service/resource/servicev2/create_test.go]: Could we have these struct members as private? A description field here would be nice too, aligning with `desired_test`

File: service/resource/servicev2/create_test.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+		resourceConfig.K8sClient = fake.NewSimpleClientset()
+		resourceConfig.Logger = microloggertest.New()
+		newResource, err = New(resourceConfig)
+		if err != nil {
+			t.Fatal("expected", nil, "got", err)
+		}
+	}
+
+	for i, tc := range testCases {
+		result, err := newResource.newCreateChange(context.TODO(), tc.Obj, tc.Cur, tc.Des)
```
Comment (fgimenez) [inline service/resource/servicev2/create_test.go]: Also like `desired_test`, it would be nice to have subtests here.

File: service/resource/servicev2/create_test.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+		if err != nil {
+			t.Fatal("case", i+1, "expected", nil, "got", err)
+		}
+		if tc.ExpectedService == nil {
+			if tc.ExpectedService != result {
+				t.Fatal("case", i+1, "expected", tc.ExpectedService, "got", result)
+			}
+		} else {
+			name := result.(*apiv1.Service).Name
```
Comment (fgimenez) [inline service/resource/servicev2/create_test.go]: Could we check the type assertion here?

File: service/resource/servicev2/delete_test.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+	"github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/micrologger/microloggertest"
+	apismetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+	apiv1 "k8s.io/client-go/pkg/api/v1"
+)
+
+func Test_Resource_Service_newDeleteChange(t *testing.T) {
+	testCases := []struct {
+		Obj             interface{}
```
Comment (fgimenez) [inline service/resource/servicev2/delete_test.go]: Same comments as in `create_test`

File: service/resource/servicev2/delete.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+	desiredService, err := toService(desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "debug", "finding out if the service has to be deleted")
+
+	var serviceToDelete *apiv1.Service
+	if currentService != nil {
+		serviceToDelete = desiredService
```
Comment (fgimenez) [inline service/resource/servicev2/delete.go]: Shouldn't we delete the current service?

File: service/resource/servicev2/delete_test.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+					Name: "al9qy",
+				},
+			},
+			Des: &apiv1.Service{
+				TypeMeta: apismetav1.TypeMeta{
+					Kind:       "Service",
+					APIVersion: "v1",
+				},
+				ObjectMeta: apismetav1.ObjectMeta{
+					Name: "al9qy",
```
Comment (fgimenez) [inline service/resource/servicev2/delete_test.go]: Maybe another test case with different and not nil current and desired state would be good.

File: service/resource/servicev2/create_test.go (hunk 2 above)
Comment (rossf7) [reply service/resource/servicev2/create_test.go]: Good point, fixed. These were copy paste from another resource.

File: service/resource/servicev2/create_test.go (hunk 3 above)
Comment (rossf7) [reply service/resource/servicev2/create_test.go]: Done and yes I think we should use subtests in more projects.

File: service/resource/servicev2/delete_test.go (hunk 7 above)
Comment (rossf7) [reply service/resource/servicev2/delete_test.go]: Good point, done.

File: service/resource/servicev2/delete.go (hunk 6 above)
Comment (rossf7) [reply service/resource/servicev2/delete.go]: I think it should be the desired state but I changed the checks in newCreateChange and newDeleteChange to also check the service name.

**Comment by rossf7** [comment]
@fgimenez Changes made. PTAL
